packages/doenetml-prototype/src/renderers/components/jsxgraph/listeners.ts

```diff
@@ -1,56 +1,143 @@
     const interactionState: InteractionState = {
         dragActive: false,
         pointerAtDown: [0, 0],
         pointerIsDown: false,
     };
 
-    obj.on("drag", function (_e) {
-        const e = _e as MouseEvent;
-        const viaPointer = e.type === "pointermove";
-
-        // Protect against very small unintended drags
-        if (
-            !viaPointer ||
-            Math.abs(e.x - interactionState.pointerAtDown[0]) > 0.1 ||
-            Math.abs(e.y - interactionState.pointerAtDown[1]) > 0.1
-        ) {
-            interactionState.dragActive = true;
-        }
-        console.log("dragging", obj.elType, { interactionState, obj });
-    });
-    obj.on("down", function (_e) {
-        const e = _e as MouseEvent;
-        interactionState.pointerAtDown = [e.x, e.y];
-        console.log("down", obj.elType, { interactionState, obj });
-    });
-    obj.on("hit", function (_e) {
-        interactionState.dragActive = false;
-        console.log("hit", obj.elType, { interactionState, obj });
-    });
-    obj.on("up", function (_e) {
-        interactionState.dragActive = false;
-        console.log("up", obj.elType, { interactionState, obj });
-    });
-    obj.on("keyfocusout", function (_e) {
-        interactionState.dragActive = false;
-        console.log("keyfocusout", obj.elType, { interactionState, obj });
-    });
-    obj.on("keydown", function (_e) {
-        const e = _e as KeyboardEvent;
-        if (e.key === "Enter") {
+    const listenersAdded: GraphListeners = {};
+
+    if (listenerActions.drag) {
+        const dragListener = function (_e: Event) {
+            const e = _e as MouseEvent;
+            const viaPointer = e.type === "pointermove";
+
+            // Protect against very small unintended drags
+            if (
+                !viaPointer ||
+                Math.abs(e.x - interactionState.pointerAtDown[0]) > 0.1 ||
+                Math.abs(e.y - interactionState.pointerAtDown[1]) > 0.1
+            ) {
+                interactionState.dragActive = true;
+            }
+            listenerActions.drag?.(e, interactionState);
+        };
+
+        obj.on("drag", dragListener);
+        listenersAdded.drag = dragListener;
+    }
+
+    if (listenerActions.down) {
+        const downListener = function (_e: Event) {
+            const e = _e as MouseEvent;
+            interactionState.pointerAtDown = [e.x, e.y];
+            interactionState.pointerIsDown = true;
+            listenerActions.down?.(e, interactionState);
+        };
+        obj.on("down", downListener);
+        listenersAdded.down = downListener;
+    }
+
+    if (listenerActions.up) {
+        const upListener = function (_e: Event) {
+            const e = _e as MouseEvent;
             interactionState.dragActive = false;
-        }
-        console.log("keydown", obj.elType, { interactionState, obj });
-    });
+            interactionState.pointerIsDown = false;
+            listenerActions.up?.(e, interactionState);
```

Comment:
Maybe we want to save `const up = listenerActions.up;` and then use `up` directly here. It would never be null and wouldn't change if `listenerActions` changed.

Reply:
Ah, that's better. I was annoyed that I had to use the `?`.